pkg/controller/build/build_controller.go

```diff
@@ -235,67 +233,60 @@ func (r *ReconcileBuild) Reconcile(request reconcile.Request) (reconcile.Result,
 	}
 
 	// Validate if the referenced secrets exist in the namespace
-	var secretNames []string
+	secretRefMap := map[string]build.BuildReason{}
 	if b.Spec.Output.SecretRef != nil && b.Spec.Output.SecretRef.Name != "" {
-		secretNames = append(secretNames, b.Spec.Output.SecretRef.Name)
+		secretRefMap[b.Spec.Output.SecretRef.Name] = build.SpecOutputSecretRefNotFound
 	}
 	if b.Spec.Source.SecretRef != nil && b.Spec.Source.SecretRef.Name != "" {
-		secretNames = append(secretNames, b.Spec.Source.SecretRef.Name)
+		secretRefMap[b.Spec.Source.SecretRef.Name] = build.SpecSourceSecretRefNotFound
 	}
 	if b.Spec.BuilderImage != nil && b.Spec.BuilderImage.SecretRef != nil && b.Spec.BuilderImage.SecretRef.Name != "" {
-		secretNames = append(secretNames, b.Spec.BuilderImage.SecretRef.Name)
+		secretRefMap[b.Spec.BuilderImage.SecretRef.Name] = build.SpecRuntimeSecretRefNotFound
 	}
 
-	if len(secretNames) > 0 {
-		if err := r.validateSecrets(ctx, secretNames, b.Namespace); err != nil {
-			b.Status.Reason = err.Error()
-			if updateErr := r.client.Status().Update(ctx, b); updateErr != nil {
-				// return an error in case of transient failure, and expect the next
-				// reconciliation to be able to update the Status of the object
-				return reconcile.Result{}, fmt.Errorf("errors: %v %v", err, updateErr)
-			}
-			// The Secret Resource watcher will Reconcile again once the missing
-			// secret is created, therefore no need to return an error and enter on an infinite
-			// reconciliation
-			return reconcile.Result{}, nil
+	// Validate if the referenced secrets exist
+	if len(secretRefMap) > 0 {
+		if err := r.validateSecrets(ctx, secretRefMap, b); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		if b.Status.Reason != build.SucceedStatus {
+			return r.UpdateBuildStatusAndRetreat(ctx, b)
 		}
 	}
 
-	// Validate if the build strategy is defined
+	// Validate if the referenced strategy exists
 	if b.Spec.StrategyRef != nil {
-		if err := r.validateStrategyRef(ctx, b.Spec.StrategyRef, b.Namespace); err != nil {
-			b.Status.Reason = err.Error()
-			updateErr := r.client.Status().Update(ctx, b)
-			return reconcile.Result{}, fmt.Errorf("errors: %v %v", err, updateErr)
+		if err := r.validateStrategyRef(ctx, b); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		if b.Status.Reason != build.SucceedStatus {
+			return r.UpdateBuildStatusAndRetreat(ctx, b)
 		}
 		ctxlog.Info(ctx, "buildStrategy found", namespace, b.Namespace, name, b.Name, "strategy", b.Spec.StrategyRef.Name)
 	}
 
-	// validate if "spec.runtime" attributes are valid
+	// Validate "spec.runtime" attributes
 	if utils.IsRuntimeDefined(b) {
-		if err := r.validateRuntime(b.Spec.Runtime); err != nil {
-			ctxlog.Error(ctx, err, "failed validating runtime attributes", "Build", b.Name)
-			b.Status.Reason = err.Error()
-			updateErr := r.client.Status().Update(ctx, b)
-			return reconcile.Result{}, fmt.Errorf("errors: %v %v", err, updateErr)
+		if r.validateRuntimeFailed(b) {
```

Comment:
attempt to better illustrate my earlier comment

Say we define a new interface with a method signature of something like

ValidateBuild(context, buildObj, mapOfAdditionalParametersSpecifcToImplementation)

And we create an array of these, associate that array with this reconciler, and then rather than call r.validateSecrets, we call ValidateBuildi instead of r.validateSecrets

in this case, mapOfAdditionalParametersSpecifcToImplementation may be nil since this implementation of ValidateBuild.  As to context, even though you currently are not supplying it to `validateRuntimeFailed` I would not be surprised if the implementation of `ValidateBuild` for this can either in fact use it, or it not, it can just ignore it.

To me, this is a stylistic / implementation detail.